### DISCUSSION TITLE
DSL Anyone?

### DISCUSSION THREAD

**Discussion by qui3xote** [custom-components/pyscript#296] [Show and tell]
After moving a few automations over to pyscript, I got inspired to try and simplify the process of writing automations even further. YAML (IMO) stinks for the job and while I love PyScript, it seems like overkill for *most* home automation jobs. So I got inspired to write a domain specific language (which I call OttoScript) on top of PyScript. OttoScript abstracts away nearly all of the technical details of working with HomeAssistant in favor of a simple, declarative language with minimal syntax rules (though lots of grammatical constraints). I've just published v0.1 and I've moved several of my simpler automations over to it and have been running for several days - so far, I'm quite pleased with it, and now that it's up and running, adding new features is proving to be pretty easy.  

Here's a simple (albeit silly) example script that demonstrates many of the working features:

```
WHEN input_boolean.working CHANGES TO 'on' 
  OR binary_sensor.office_occupied CHANGES FROM 'off' TO 'on'
  IF lights.office_lights == 'off' AND sensor.hue_motion_1_illuminance < 20
   THEN 
     DIM lights.office_lights TO 30%
 IF lights.office_lights == 'on' AND sensor.hue_motion_1_illuminance < 20
  THEN
    DIM lights.office_lights BY +30%
  IF 1 = =1 and (10 > 20 OR 30 > 40)
    THEN
      WAIT 30 SECONDS
      TURN ON switch.office_noise_machine
```

The capitalization of keywords, line breaks and indentation are all purely optional (like SQL).

While the set of supported commands is small right now, OttoScript is quite extensible and it should very easy to add more. I also intend to add more conditionals (like a CASE statement and ELSE support) and the ability to nest them, which will require a bit more work, but doesn't seem too bad at all. I would *love* feedback and contributions from this community if anyone is interested!

Installation requires cloning two repositories - the [OttoScript parser](https://github.com/qui3xote/ottoscript) and the [Pyscript App](https://github.com/qui3xote/ottopyscript). The limitations of PyScript would've made writing it as a fully native app very cumbersome, so I chose to build most of it as a separate python library, and then a thin 'interpreter' that is passed to the library for callback functions.  

EDIT: typo

**Comment by craigbarratt** [comment]
Wow - that's impressive.  Thanks for sharing!

**Reply by qui3xote** [reply]
Thanks! And thanks for pyscript - it's such a huge leap up over existing tooling. I wouldn't have even attempted OttoScript without it. I really want to make it easy for new folks to get started with home assistant, even with minimal or no programming experience. 

By the way, one issue I'm running into: once pyscript loads an external package, it **really** does not want to let it go. I often have to resort to a full HASS reboot to get the OttoScript library reloaded. Any suggestions? Debugging can get real painful when you have to reboot to get changes reflected....

**Reply by dlashua** [reply]
It's a python thing. I've not tried it with PyScript, but, with plain old python, this is the trick II use to reload a library inside of a running python process.

https://www.geeksforgeeks.org/reloading-modules-python/

**Reply by qui3xote** [reply]
Yeah, that's what I've been using - it seems to work _some_ of the time with PyScript, and never while using the jupyter kernel. Haven't figured out a pattern to when it does or doesn't work running live. I'm guessing something about the way PyScript wraps everything up in evalfuncs and/or the async evaluation is keeping it from happening in the right namespace? I'm way out of my depth on that kind thing, though.
